packages/ui/src/ui.tsx

```diff
@@ -0,0 +1,102 @@
+import * as ds from "@devicescript/core"
+import { Image, ImageContext } from "@devicescript/graphics"
+
+export abstract class BaseElement<T> implements JSX.Element {
+    constructor(props: T) {
+        Object.assign(this, props)
+    }
+    abstract render(ctx: ImageContext): void
+}
+
+export abstract class ParentElement<T> extends BaseElement<T> {
+    children?: JSX.Element | JSX.Element[]
+
+    renderChildren(ctx: ImageContext) {
+        if (!this.children) return
+        if (Array.isArray(this.children))
+            for (const ch of this.children)
+                ch.render(ctx)
+        else
+            this.children.render(ctx)
+    }
+}
+
+export class Translate extends ParentElement<{ x: number, y: number } & JSX.BaseProps> {
+    x: number
+    y: number
+
+    render(ctx: ImageContext): void {
+        ctx.save()
+        ctx.translate(this.x, this.y)
+        this.renderChildren(ctx)
+        ctx.restore()
+    }
+}
+
+export class Text extends BaseElement<{ children: string }> {
+    children: string
+
+    render(ctx: ImageContext): void {
+        ctx.fillText(this.children, 0, 0)
+    }
+}
+
+export class Rect extends BaseElement<{ w: number, h: number }> {
+    w: number
+    h: number
+
+    render(ctx: ImageContext): void {
+        ctx.strokeRect(0, 0, this.w, this.h)
+    }
+}
+
+export class FillRect extends Rect {
+    override render(ctx: ImageContext): void {
+        ctx.fillRect(0, 0, this.w, this.h)
+    }
+}
+
+export class Fragment extends ParentElement<JSX.BaseProps> {
+    render(ctx: ImageContext): void {
+        this.renderChildren(ctx)
+    }
+}
+
+
+export function HorizontalGauge(props: { w: number, h: number, used: number }): JSX.Element {
+    const { w, h, used } = props
```

Comment:
react typically does not use acornyms, so w -> width, h -> height will feel more natural